src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -346,30 +384,171 @@ public static string GetAtomicGuardKey(string docId)
+            BlittableJsonReaderObject command,
+            string id,
+            int shardIndex,
+            long currentCommandCount,
+            (long, string)? clusterGuardAddition)
+        {
+            if (command.TryGet(nameof(ClusterTransactionDataCommand.Type), out string type) == false)
+                throw new InvalidOperationException($"Got command with no type defined: {command}");
+
+            var result = new DynamicJsonValue
+            {
+                [nameof(ICommandData.Type)] = type,
+                [Constants.Documents.Metadata.LastModified] = DateTime.UtcNow,
+            };
+            
+            switch (type)
             {
-                var count = commandsCountPerDatabase.ReadInt64(databaseSlice) ?? 0;
-                using (GetPrefix(context, DatabaseName, out var prefixSlice, count))
+                case nameof(CommandType.PUT):
+                    if (command.TryGet(nameof(ClusterTransactionDataCommand.Document), out BlittableJsonReaderObject document)
+                        && document.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata)
+                        && metadata.TryGet(Constants.Documents.Metadata.Flags, out DocumentFlags flags))
+                    {
+                        result[Constants.Documents.Metadata.Flags] = flags | DocumentFlags.FromClusterTransaction;
+                    }
+                    result[Constants.Documents.Metadata.Id] = id;
+                    break;
+                case nameof(CommandType.DELETE):
+                    result[Constants.Documents.Metadata.IdProperty] = id;
+                break;
+                    default:
+                        throw new InvalidOperationException(
+                            $"Database cluster transaction command type can be {CommandType.PUT} or {CommandType.PUT} but got {type}");
+            }
+
+            var databaseId = rawRecord.Shards[shardIndex].DatabaseTopologyIdBase64;
+            var changeVector = BatchHandler.ClusterTransactionMergedCommand.GetClusterWideChangeVector(databaseId, currentCommandCount, clusterGuardAddition);
+
+            result[Constants.Documents.Metadata.ChangeVector] = changeVector;
+            return result;
+        }
+
+        private void InsertCommandBatch(ClusterOperationContext context, long index, string databaseName, Tree commandsCountPerDatabase, Table items,
+            BlittableJsonReaderObject commands, long commandsCount)
+        {
+            SaveCommandBatch(context, index, databaseName, commandsCountPerDatabase, items, commands, commandsCount, ItemsInsert);
+        }
+        private void SetCommandBatch(ClusterOperationContext context, long index, string databaseName, Tree commandsCountPerDatabase, Table items,
+            BlittableJsonReaderObject commands, long commandsCount)
+        {
+            SaveCommandBatch(context, index, databaseName, commandsCountPerDatabase, items, commands, commandsCount, ItemsSet);
```

Comment:
When sending a database request we send the last known raft command index (saved in the document store level).
Sending cluster-wide transactions with a write request to let say only $0 will update the client with the new raft index of the cluster transaction command.
If then the client sends another database request (line `Load`) it will contain the new index.
If that database request is then directed to the same shard the directed request will contain the new index to wait for that index to be applied on the shard so the client will not get an old result.
But if the `Load` request is directed to another shard (let's say $1) it will wait for the index to be updated but the origin command is not relevant for him.
This is only relevant if another raft command didn't apply on that shard (like creating/modifying index) and updated the index to be higher than the cluster transaction command index.
I included here a way to solve it.
The cluster state machine will send empty command if no database command is related to it.
That will throw an exception in the previous implementation because the `prevCount` is the same so the key is the same and we use `Inset` (NewOnly).
I know @karmeli87 like "guard" like that to catch potential bugs (And I like the idea) so I modified only the shard implementation to use `set`.

Reply:
ping me about that I want to better understand this

Reply:
The database _should_ update it for newer values, yes.